src/include/mpir_request.h

```diff
@@ -204,11 +204,59 @@ struct MPIR_Request {
 #endif
 };
 
+/* Multiple Request Pools
+ * Request objects creation and freeing is in a hot path. Multiple pools allow different
+ * threads to access different pools without incurring global locks. Because only request
+ * objects benefit from this multi-pool scheme, the normal handlemem macros and functions
+ * are extended here for request objects. It is separate from the other objects, and the
+ * bit patterns for POOL and BLOCK sizes can be adjusted if necessary.
+ *
+ * NOTE: MPIR_Request_create/free are patched to work with pools.
+ */
+/* Handle Bits - 2+4+6+8+12 - Type, Kind, Pool_idx, Block_idx, Object_idx */
+#define REQUEST_POOL_MASK    0x03f00000
+#define REQUEST_POOL_SHIFT   20
+#define REQUEST_POOL_MAX     64
+#define REQUEST_BLOCK_MASK   0x000ff000
+#define REQUEST_BLOCK_SHIFT  12
+#define REQUEST_BLOCK_MAX    256
+#define REQUEST_OBJECT_MASK  0x00000fff
+#define REQUEST_OBJECT_SHIFT 0
+#define REQUEST_OBJECT_MAX   4096
+
+#define MPIR_REQUEST_NUM_POOLS REQUEST_POOL_MAX
 #define MPIR_REQUEST_PREALLOC 8
 
-extern MPIR_Object_alloc_t MPIR_Request_mem;
-/* Preallocated request objects */
-extern MPIR_Request MPIR_Request_direct[];
+extern MPIR_Object_alloc_t MPIR_Request_mem[MPIR_REQUEST_NUM_POOLS];
+extern MPIR_Request MPIR_Request_direct[MPIR_REQUEST_NUM_POOLS][MPIR_REQUEST_PREALLOC];
```

Comment:
This will have some impact on memory footprint (8 -> 512 direct requests). We might need to be more conservative. With ch3, `MPIR_Request` size is ~1k. ch4:ofi is 328 bytes.

Reply:
I can add some code to make direct-pool only available for vci 0 -- to reduce the impact.

Reply:
This sounds reasonable to do for now.